File: Palindrome_Number.cc
/*
Determine whether an integer is a palindrome. Do this without extra space.
*/
class Solution 
{
public:
    bool isPalindrome(int x) 
    {
        if (x < 0) 
            return false;
        
        int d = 1;
        while (x / d >= 10)
            d *= 10;  // 直接到最高位数
        
        while (x > 0)
        {
            int high = x / d;  // 最高位
            int low = x % 10; // 最低位
            
            if (high != low)
                return false;
            
            x = x % d / 10; // 去掉最高位 和 最低位
            d /= 100; 
        }
        
        return true;
    }
};

File: go/palindrome-number/palindrome-number.go
package palindrome_number

/*
Determine whether an integer is a palindrome. Do this without extra space.
*/

// 回文数字，比如12321
func isPalindromeNumber(x int) bool {
	if x < 0 {
		return false
	}

	div := 1 // 数量级
	for n := x; n >= 10; n /= 10 {
		div *= 10
	}

	for x > 0 {
		if x/div != x%10 {
			return false
		}
		x = (x % div) / 10 // 去掉首尾
		div /= 100
	}

	return true
}
